### PR TITLE
Webpack all the things!

### PR DESCRIPTION
**Pull request by thogra**
Introduces [webpack](http://webpack.github.io) as the packaging mechanism for the sdk-js. The setup currently packages the sdk, and the optional spid-uri module, in these formats: 
 - As a variable  ( i.e. `var SPiD = ...`)
 - AMD
 - Common Js

Because of the number of files produced when building, the resulting files are moved from `/dist/...` to `/dist/version.number/...`

Aslo&mdash;in order to get the unit tests to work&mdash;the test setup (blanket + chai + mocha) was replaced with (karma + chai + mocha + sinon). ~~Unfortunately this means there's no code coverage tool ATM, but it should be possible to use Istanbul in the future~~ A Coverage report is now generated when running `grunt test`, although there's no automatic enforcement of line coverage %.

In addition, this PR re-introduces the `SPiD.notLoggedin`, `SPiD.statusChange` and `SPiD.sessionInit` events on `hasSession()` calls. Also: the `Spid.Cache` module is now removed from `SPiD.Persist`, and is used again with the `cache` config option in `hasSubscription/hasProduct`

### CODE REVIEW

**Comment by thogra** [comment]
Ping @joawan @erikfried

File: src/spid-sdk.js [hunk 1]
```diff
@@ -1,165 +1,171 @@
...
 
-    function hasProduct(productId, callback) {
-        var cache = this.Persist,
-            util = this.Util;
-        callback = callback || function() {};
-        if(cache) {
-            var cacheVal = cache.get('prd_{id}'.replace('{id}', productId));
-            if(cacheVal && (cacheVal.refreshed + _options.refresh_timeout) > util.now()) {
-                return callback(null, cacheVal);
+function hasSession(callback) {
```
Comment (thogra) [inline src/spid-sdk.js]: I was thinking: do we need/want some way to bypass the cache mechanism here, perhaps via an additional parameter, @joawan ?

File: Gruntfile.js [hunk 2]
```diff
@@ -33,66 +47,57 @@ module.exports = function(grunt) {
...
   grunt.loadNpmTasks('grunt-contrib-uglify');
-  grunt.loadNpmTasks('grunt-contrib-concat');
   grunt.loadNpmTasks('grunt-contrib-jshint');
-  grunt.loadNpmTasks('grunt-contrib-watch');
-  grunt.loadNpmTasks('grunt-blanket-mocha');
+  grunt.loadNpmTasks('grunt-karma');
 
   // Default task.
-  grunt.registerTask('default', ['jshint:sdk', 'jshint:tracker', 'concat', 'uglify']);
+  grunt.registerTask('default', ['jshint:sdk', 'webpack', 'uglify']);
```
Comment (thogra) [inline Gruntfile.js]: I'm considering adding the "grunt-release-it" as the way to make releases. But we could also just go with the NPM variant of release it.

File: Gruntfile.js [hunk 3]
```diff
@@ -33,66 +47,57 @@ module.exports = function(grunt) {
...
-        dest: 'dist/spid-sdk-<%= pkg.version %>.js'
-      },
-      sdktracker: {
-        src: ['<%= jshint.sdk.src %>', '<%= jshint.tracker.src %>'],
-        dest: 'dist/spid-sdk-pulse-<%= pkg.version %>.js'
       }
     },
     uglify: {
       options: {
         banner: '<%= banner %>'
```
Comment (thogra) [inline Gruntfile.js]: Note to self: this is currently broken

File: src/spid-sdk.js (hunk 1 above)
Comment (thogra) [inline src/spid-sdk.js]: Hmm, what'd be the use case for such a "force" parameter? Do you know of a concrete example where this'd be useful (as opposed to just initialising the SDK with the ´storage´ option set to `null`) @erikfried ?

Comment (erikfried) [inline src/spid-sdk.js]: Our concrete example is that we force reload on first request following an oauth roundtrip to spid. We figure it is extremely likely that the users state is somehow updated then, e g name updated, product purchased, user logged in/out.

Comment (joawan) [inline src/spid-sdk.js]: I have something against boolean parameters to function calls, as cite uncle Bob "it's a code smell".
I propose this to be a little more verbose. First, do a call to clear cache, and then to a call to hasSession. Thoughts?

Comment (thogra) [inline src/spid-sdk.js]: Conclusion: we'll expose the `persist` module (as `SPiD.persist` I guess)

File: Gruntfile.js [hunk 4]
```diff
@@ -33,66 +47,61 @@ module.exports = function(grunt) {
...
+      sdkAmd: {
+        src: 'dist/spid-sdk-amd.js',
+        dest: 'dist/spid-sdk-amd.min.js'
       },
-      sdktracker: {
-        src: '<%= concat.sdktracker.dest %>',
-        dest: 'dist/spid-sdk-pulse-<%= pkg.version %>.min.js'
+      sdkCommonJs: {
+        src: 'dist/spid-sdk-commonjs2.js',
+        dest: 'dist/spid-sdk-commonjs2.min.js'
```
Comment (joawan) [inline Gruntfile.js]: Dist files need to be versioned

Comment (joawan) [inline Gruntfile.js]: As in, version number in file name.

File: src/spid-event.js [hunk 5]
```diff
@@ -1,83 +1,82 @@
-/*global SPiD:false*/
-;(function(exports) {
+/*global require:false, module:false*/
 
-    var _subscribers = {};
+var _subscribers = {},
```
Comment (joawan) [inline src/spid-event.js]: rename var, remove underscore

**Pull request merged**